### PR TITLE
Basic user management on the CLI

### PR DESCRIPTION
**Pull request by drubin**
This manages users in the most basic form. It currently supports `createUser` `updateUser` but these are isolated and on their own.

This is mostly to get high-level feedback about the configuration setup, because this is the first command that has to use an explicit connection to  a db I added it to the config file.

## Example config

```yaml
connection:
  uri: mongodb://localhost:27017/admin
users:
- user: 'david'
  password: 'pass123'
  customData:
    foo: bar
```

## Example Output

```
Setting up users
{ _: [ 'users' ],
  f: 'hydra.yaml',
  connection: { uri: 'mongodb://localhost:27017/admin' },
  users:
   [ { user: 'david', password: 'pass123', customData: [Object] },
     { user: 'immanuel', password: 'pass123', customData: [Object] } ],
  '$0': 'hydra' }
creating user { user: 'david', password: 'pass123', customData: { foo: 'bar' } }
User exists updating instead
{ ok: 1 }
creating user { user: 'immanuel', password: 'pass123', customData: { foo: 'bar' } }
User exists updating instead
{ ok: 1 }
```

We need to think about how to wrap this logic into a bit more abstract lib (similar to how we did replication), because obviously this logic doesn't belong in the CLI layer but it was just a quick hack to prove it.

### CODE REVIEW

File: examples/users/hydra.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+connection:
+  uri: mongodb://localhost:27017/admin
+users:
+- user: 'david'
+  pwd: 'pass123'
```
Comment (pelzerim) [inline examples/users/hydra.yaml:5]: How about making this `password` instead? `pwd` requires the user to remember this abbreviation. See https://github.com/ryanmcdermott/clean-code-javascript#use-meaningful-and-pronounceable-variable-names

File: examples/users/hydra.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+connection:
```
Comment (pelzerim) [inline examples/users/hydra.yaml:1]: I like this. Original hydra (basically) just had a root level `uri`, but this is more extedable

File: lib/cli/user.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+exports.handler = async (argv) => {
+  logger.debug('Setting up users')
+  logger.debug(argv)
+
+  const { uri } = argv.connection
+  const { users } = argv
+  const client = new MongoClient(uri)
+  await client.connect()
+  const userCommands = new UserCommands(client)
```
Comment (pelzerim) [inline lib/cli/user.js:27]: (This is obviously wip, but leaving a comment to be thorough)

This should be a `UserState` instead that has a `apply` function and wraps this

File: lib/cli/user.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+  for (let i = 0; i < users.length; i++) {
+    const user = users[i]
+    logger.debug('creating user', user)
+    try {
```
Comment (pelzerim) [inline lib/cli/user.js:32]: I think this should "get" the user before creating it instead:

- Get user
- If user exists: update user 
- if user does not exist: create user

File: lib/core/state/mongo-user.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+class MongoUser {
```
Comment (pelzerim) [inline lib/core/state/mongo-user.js:1]:
```suggestion
class MongoUserState {
```

to be consistent

File: examples/users/hydra.yaml (hunk 1 above)
Comment (drubin) [reply examples/users/hydra.yaml:5]: It's pwd in mongo. So we have the question here to follow their naming or use better naming. In general I feel we should try use upstream terminologies as much as possible.

But this annoyed me too to be fair I originally had it as password and changed it. 

Let me know your thoughts.

File: lib/cli/user.js (hunk 3 above)
Comment (drubin) [reply lib/cli/user.js:27]: Agreed this was just a hack to prove it and make it work and get feedback on naming and structure.

File: lib/cli/user.js (hunk 4 above)
Comment (drubin) [reply lib/cli/user.js:32]: When I move this into user state I will do this too.

**Comment by drubin** [comment]
I will rebase the commits before merging, but this is the 2nd iteration of basic user management. I am not going to close the ticket yet because it doesn't yes support edit/removing/adding (maybe we should postpone this till after sharding as its actually complex to get right and there are better tools for this.

File: lib/cli/user.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  logger.debug('Setting up users')
+
+  const { uri } = argv.connection
+  const { users } = argv
+  const client = new MongoClient(uri)
+  await client.connect()
+
+  const userState = new MongoUsersState({ users, client })
+  const response = await userState.sync()
+  logger.debug(response)
```
Comment (pelzerim) [inline lib/cli/user.js:22]: No need for this, right?

File: lib/core/commands/user-commands.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    const command = {
+      updateUser: user.user,
+      pwd: user.pwd,
+      customData: user.customData || {},
+      roles: user.roles || []
+    }
+    return await this.client.command(command)
+  }
+
+  // https://docs.mongodb.com/manual/reference/command/// https://docs.mongodb.com/manual/reference/command/updateUser/
```
Comment (pelzerim) [inline lib/core/commands/user-commands.js:28]: Broken link

File: lib/core/state/mongo-users-state.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+   * @param {String} settings.users users to create
+   */
+  constructor(settings) {
+    const { connection } = settings
+    this.users = settings.users
+    // Client
+    this.client = settings.client || new MongoClient(connection.uri)
+  }
+
+  async sync() {
```
Comment (pelzerim) [inline lib/core/state/mongo-users-state.js:18]: The current convention is „apply“

File: lib/core/state/mongo-users-state.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    for (let i = 0; i < this.users.length; i += 1) {
+      const user = this.users[i]
+      logger.debug('checking if user exits', user.user)
+      const response = await userCommands.usersInfo(user.user)
+      logger.debug('user resp', response)
+      // If this array is filled it means the user exists
+      if (response.users.length) {
+        logger.debug('updating user', user)
+        const resp = await userCommands.updateUser(user)
+        responses.push(resp)
```
Comment (pelzerim) [inline lib/core/state/mongo-users-state.js:33]: Hydra should deal with responses, this currently fully swallows errors. See https://github.com/unumotors/mongo-hydra/blob/9ae392709da9f992abd1fb496a0180d6dab14490/lib/core/state/replication.js#L31

File: lib/core/state/mongo-users-state.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    // for now it only creates/updates new ones
+    const userCommands = new UserCommands(this.client)
+
+    for (let i = 0; i < this.users.length; i += 1) {
+      const user = this.users[i]
+      logger.debug('checking if user exits', user.user)
+      const response = await userCommands.usersInfo(user.user)
+      logger.debug('user resp', response)
+      // If this array is filled it means the user exists
+      if (response.users.length) {
```
Comment (pelzerim) [inline lib/core/state/mongo-users-state.js:30]: This can be an early return „if(!response.users.length“

Comment (drubin) [reply lib/core/state/mongo-users-state.js:30]: How would that work? This is per user.

Comment (pelzerim) [reply lib/core/state/mongo-users-state.js:30]: Ah yes, I see.

**Comment by pelzerim** [comment]
closes #46

File: lib/core/state/mongo-users-state.js (hunk 9 above)
Comment (drubin) [reply lib/core/state/mongo-users-state.js:33]: This doesn't swallow them it returns them. I am happy to switch to exception based handling but  this has ability to show which users were successful.

The problem with exception based flows is we can't automatically disconnect easily...

**Comment by drubin** [comment]
I will fix the issues listwd above but are there any major issues that we would need to fix before merging?

The above things seem pretty minor.

File: lib/core/state/mongo-users-state.js (hunk 9 above)
Comment (pelzerim) [reply lib/core/state/mongo-users-state.js:33]: The issue is currently see is that it "looks" like everything is wrong as there will just be a { ok: 0 } in a bunch of code

I'd rather fail hard here and make the user notice super hard that something went wrong?

**Comment by pelzerim** [comment]
> I will fix the issues listwd above but are there any major issues that we would need to fix before merging?
> 
> The above things seem pretty minor.

I fully agree, that everything is minor but the error handling. But tbh we can patch that in later. approved

**Comment by drubin** [comment]
I will fix them before merging

**Comment by drubin** [comment]
We should add a issue to figure out proper error handling before it gets worse.... I think we need a solid pattern before copying it around too much.

File: lib/core/state/mongo-users-state.js (hunk 9 above)
Comment (drubin) [reply lib/core/state/mongo-users-state.js]: Done

File: lib/cli/user.js (hunk 6 above)
Comment (drubin) [reply lib/cli/user.js]: Removed

File: lib/core/commands/user-commands.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
+class UserCommands {
+  constructor(client) {
+    this.client = client
+  }
+
+  // https://docs.mongodb.com/manual/reference/command/createUser/
+  async createUser(user) {
+    const command = {
+      createUser: user.user,
+
```
Comment (pelzerim) [inline lib/core/commands/user-commands.js:10]:
```suggestion
```

File: lib/core/state/mongo-users-state.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    // Client
+    this.client = settings.client || new MongoClient(connection.uri)
+  }
+
+  async apply() {
+    // this should sync the users
+    // for now it only creates/updates new ones
+    const userCommands = new UserCommands(this.client)
+
+    for (let i = 0; i < this.users.length; i += 1) {
```
Comment (pelzerim) [inline lib/core/state/mongo-users-state.js:23]: You can use `for of` here for more readability https://github.com/unumotors/mongo-hydra/blob/master/lib/core/state/replication.js#L58

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

(not a blocker)

File: lib/core/state/mongo-users-state.js [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+      const user = this.users[i]
+      logger.debug('checking if user exits', user.user)
+      const response = await userCommands.usersInfo(user.user)
+      logger.debug('user resp', response)
```
Comment (pelzerim) [inline lib/core/state/mongo-users-state.js:27]:
```suggestion
```

Might be sensitive and currently the logger logs everything

File: lib/core/state/mongo-users-state.js [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+      // If this array is filled it means the user exists
+      if (response.users.length) {
+        logger.debug('updating user', user)
```
Comment (pelzerim) [inline lib/core/state/mongo-users-state.js:30]: This prints the password in the current implementation

File: lib/core/state/mongo-users-state.js [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+        const resp = await userCommands.updateUser(user)
+        if (!resp.ok) {
+          throw new Error(`unable to update user: ${user.user}`)
```
Comment (pelzerim) [inline lib/core/state/mongo-users-state.js:33]:
```suggestion
          throw new Error(`Unable to update user: ${user.user}`)
```

File: lib/core/state/mongo-users-state.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        }
+      } else {
+        logger.debug('creating user', user)
```
Comment (pelzerim) [inline lib/core/state/mongo-users-state.js:36]:
```suggestion
        logger.debug('creating user', user.user)
```

File: lib/core/state/mongo-users-state.js (hunk 13 above)
Comment (drubin) [reply lib/core/state/mongo-users-state.js]: It doesn't but I removed it anyway. :)

File: lib/core/state/mongo-users-state.js (hunk 12 above)
Comment (drubin) [reply lib/core/state/mongo-users-state.js]: Done, I always get confused with the `for in` that can only be used with objects, thanks for pointing this out!

**Comment by drubin** [comment]
I am going to merge this as this has been dragging on too long and all the issues were cosmetic.

**Pull request merged**